msteams-platform/sbs-botbuilder-conversation-AI.yml

```diff
@@ -0,0 +1,709 @@
+
+
+        1. Add the following code to define storage and application:
+
+            ```typescript
+            // Define storage and application
+            const storage = new MemoryStorage();
+            const app = new Application<ApplicationTurnState>({
+                storage,
+                ai: {
+                    planner,
+                    moderator,
+                    promptManager,
+                    prompt: 'chat',
+                    history: {
+                        assistantHistoryType: 'text'
+                    }
+                }
+            });
+            ```
+
+            * The `Application` class replaces the Teams Activity Handler class from Bot Framework SDK. The application class has all the information and bot logic required for an app. You can register actions or activity handlers for the app in this class.
+            * The `MemoryStorage` class is imported from the Bot Builder SDK and has a predefined logic to handle any issues when two or more events exist at the same time so that you don't want to worry about handling storage issues and focus on building your application.
+
+
+            1. Under **Import required bot service**, add `MemoryStorage`.
+
+        1.  Add the following code to define actions:
+
+            ```typescript
+            app.ai.action(AI.FlaggedInputActionName, async (context, state, data) => {
+                await context.sendActivity(`I'm sorry your message was flagged: ${JSON.stringify(data)}`);
+                return false;
+            });
+
+            app.ai.action(AI.FlaggedOutputActionName, async (context, state, data) => {
+                await context.sendActivity(`I'm not allowed to talk about such things.`);
+                return false;
+            });
+            ```
+            
+            When the `OpenAIModerator` reviews and identifies a flag in an input or output, it triggers the action `AI.FlaggedInputActionName` or `AI.FlaggedOutputActionName` and the `context.sendActivity` sends a message to the user. If the user receives a message for an flagged  input or output, the user can control what he wants to do.
+
+        1. Add the following code to define history:
+
+            ```typescript
+            app.message('/history', async (context, state) => {
+            const history = ConversationHistory.toString(state, 2000, '\n\n');
+            await context.sendActivity(history);
+            });
+            ```
+
+           The `ConversationHistory` class remembers context across messages. It automatically populates the {{$history}} variable available in the `prompt.txt` file.
+
+        1. By default Bot builder uses `bot` as a variable, since we are constructing an app for the `new Application`. We need to replace `await bot.run(context) with `await app.run(context)`.
```

Comment:
<img width="457" alt="image" src="https://github.com/MicrosoftDocs/msteams-docs/assets/115390646/e645af44-0c72-4f95-841a-bdb463ca2272">
There's a tick (`) mismatch

Reply:
updated the code tick.